compiler/semcall.nim

```diff
@@ -505,7 +505,15 @@ proc explicitGenericSym(c: PContext, n: PNode, s: PSym): PNode =
 
   for i in 1..sonsLen(n)-1:
     let formal = s.ast.sons[genericParamsPos].sons[i-1].typ
-    let arg = n[i].typ
+    var arg = n[i].typ
+    # try transforming the argument into a static one before feeding it into
+    # typeRel
+    if formal.kind == tyStatic and arg.kind != tyStatic:
+      let evaluated = c.semTryConstExpr(c, n[i])
+      if evaluated != nil:
+        arg = newTypeS(tyStatic, c)
+        arg.sons = @[evaluated.typ]
+        arg.n = evaluated
     let tm = typeRel(m, formal, arg)
```

Comment:
I think the problem here was that `typeRel` was used directly, instead of using the full `paramTypesMatchAux`, which has grown with many additional bits of logic over the years. Basically, in this fix you've extracted a little bit of the code in `paramTypesMatchAux` here.

Reply:
Yup, I did not know if using the full-blown matching routine had any unwanted side effect, feel free to change this to use it.

Implicit `static`s may help remove many of those special-cases.